### ISSUE TITLE
Add build-system.requires to lock file

### ISSUE THREAD

**Issue by adisbladis** [python-poetry/poetry#8752]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [FAQ](https://python-poetry.org/docs/faq/) and general [documentation](https://python-poetry.org/docs/) and believe that my question is not already covered.

### Problem

A notable thing missing from `poetry.lock` are build-systems used to build from sdists.

This is useful in efforts like [poetry2nix](https://github.com/nix-community/poetry2nix) that uses Poetry metadata.
Nix build environments are constructed in a much stricter way where we cannot dynamically install/load build-systems.
Currently users of this tool have to resort to manually adding overrides into Nix code to work around this missing metadata.

This issue has been reported many times over the years in various forms:
- https://github.com/nix-community/poetry2nix/issues/568
- https://github.com/nix-community/poetry2nix/issues/1435

And has generally been a large source of maintenance that could be completely automated away.

Adding this metadata also has other uses such as building tools that can statically inspect the build-time dependency graph.

### Solution

One more possible entries in `poetry.lock`:
- `build-system.requires`

We only need to know `build-system.requires` & up front when constructing the build environment.
There is no need to add `build-system.backend` or `build-system.backend-path`.

The PyPi API does not have this metadata. It would need to be extracted from the sdist.

### Note

Do note that this is _not_ a suggestion to lock build systems, merely to add the contents of `build-system.requires`.

**Comment by dimbleby** [comment]
additionally poetry mostly does not care about build requirements when building the lock file - because all of the metadata that it needs is available either directly from a server API, or in an already-built wheel.

ie obtaining this information would hurt every other user: by forcing poetry to download and examine sdists during locking, which it otherwise only rarely needs to to do

**Comment by blaggacao** [comment]
(Somewhat innocently) quoting @Secrus 

> Specifying build dependencies is a concern for project maintainers, not end users.

Wouldn't the lock file be a project maintainer's truest (even: only) source of the metadata she requires?

**Comment by dimbleby** [comment]
There is some possible ambiguity here so let's be sure we're on the same page.  

I don't think you are asking for build requirements from the "current" project's `pyproject.toml` to be duplicated into `poetry.lock`: that would obviously be pointless, because you can instead simply read them directly from `pyproject.toml` - right?

Rather you are asking for all of the build requirements of all of the dependencies of the current project to be copied into `poetry.lock` - yes?

As I say, this would at the very least require poetry to do a whole pile of work - and downloading - during locking that it otherwise has no interest in doing.  Poetry just does not need this information in its lockfile: whenever poetry wants the build requirements from an sdist, it already has the sdist in hand and can read them directly.

I don't know why this isn't also true for poetry2nix but - in the nicest possible way - that sounds like a poetry2nix problem.

Perhaps you would benefit from build requirements being added to the package metadata in general: if so, you could start a conversation with the PyPA and try to get that agreed.

**Comment by dimbleby** [comment]
Really?

Well this isnt going to happen then, because it is obviously pointless!

`poetry.lock` is already insufficient without `pyproject.toml` for all sorts of reasons.  What are the circumstances in which you have `poetry.lock` for a project but not the corresponding `pyproject.toml`?  How does this fit with the second sentence of the poetry2nix readme?

> It does so by parsing pyproject.toml and poetry.lock ...

In short: just parse `pyproject.toml` in the first place!

**Comment by blaggacao** [comment]
> I don't know why this isn't also true for poetry2nix but - in the nicest possible way - that sounds like a poetry2nix problem.

While this _seems_ to be the case, `poetry2nix` &mdash; being a build toolchain &mdash; just consumes the maintainer's specced out metadata.

If the spec is incomplete, that unsurprisingly causes the issues described on the initial post above.

> Perhaps you would benefit from build requirements being added to the package metadata in general: if so, you could start a conversation with the PyPA and try to get that agreed.

That sounds actually strategically reasonable. We'd be left to discuss wether poetry may be a stepping stone in that strategy in some capacity [alongside pdm](https://github.com/pdm-project/pdm/issues/2465). I don't know enough to see what's overall most practical / pareto efficient.

**Comment by adisbladis** [comment]
> What are the circumstances in which you have poetry.lock for a project but not the corresponding pyproject.toml?

Here is the case stated as simply as possible:
You want to build a project. This project contains `pyproject.toml` & `poetry.lock`.
We can read those just fine.

We can't recurse into dependencies `pyproject.toml` files or any other source files to reason about _their_ build time dependencies statically as that requires network side effects.

**Comment by blaggacao** [comment]
> In short: just parse pyproject.toml in the first place!

It boils down to the question of "who should make the inventory?" - the locking tool or the build system?

Some (novel) build systems are so side-effect free that they only can consume an already materialized inventory during planning the build (as @adisbladis mentions: no effectful recursion onto the network possible &mdash; mostly it's part of the security model). Nix is one such build system.

**Comment by dimbleby** [comment]
When I tried to clarify whether you cared about the build requirements of the current project or of its dependencies, you said that requirements of the current project was what you wanted, and definitely not build requirements of the current project's own dependencies.

When I said that this was pointless, you said that the challenge was that "we can't recurse into dependencies pyproject.toml files or any other source files to reason about their build time dependencies..."

So I still don't understand the ask.  Do you want the lockfile to contain build requirements only of the current project, or of all of its dependencies?

However, I suspect we are only deciding what reason to give for not doing this.  It is _either_ pointless or impractical, depending which you want.

**Comment by blaggacao** [comment]
> Do you want the lockfile to contain build requirements only of the current project, or of all of its dependencies?

Clearly there was a misunderstanding. I almost saw this coming:

The ask is for the lock file to contain the build requirements of dependencies (by name) only.

> However, I suspect we are only deciding what reason to give for not doing this. It is either pointless or impractical, depending which you want.

Impractical, in this case, in a different reference frame just turns out to be a local maximum.

I wonder if there can be a route forward to the tune of? (Nobody seemed to have an opinion on that venue, yet)

> > Perhaps you would benefit from build requirements being added to the package metadata in general: if so, you could start a conversation with the PyPA and try to get that agreed.

> That sounds actually strategically reasonable. We'd be left to discuss wether poetry may be a stepping stone in that strategy in some capacity alongside pdm. I don't know enough to see what's overall most practical / pareto efficient.

**Comment by dimbleby** [comment]
I cannot make sense of that last update.

I note that pdm maintainer is taking a similar view - https://github.com/pdm-project/pdm/issues/2465#issuecomment-1845063798 - that this would "tremendously affect locking performance" and is better handled by PEPs / ecosystem support.

Which is pretty much where I am too: poetry is not going to cripple its performance just to add data to the lockfile that it does not need.

**Comment by blaggacao** [comment]
Yep, I agree that addressing the challenge from this small-iterative-steps angle isn't going to go anywhere.

With the performance penalty, interests just seem not to be really aligning well in this particular instance.

The resources for addressing it via PEPs or any other lengthy ecosystem process may just not be there at this point.

However, full source reproducibility may eventually become a trending interest also for the python ecosystem. :shrug:

**Comment by dimbleby** [comment]
Also, duplicate #8216, #8261

**Comment by charmoniumQ** [comment]
> > you are asking for all of the build requirements of all of the dependencies of the current project to be copied into `poetry.`lock - yes?
> >
> > As I say, this would at the very least require poetry to do a whole pile of work - and downloading - during locking that it otherwise has no interest in doing. Poetry just does not need this information in its lockfile: whenever poetry wants the build requirements from an sdist, it already has the sdist in hand and can read them directly.
>
> poetry is not going to cripple its performance just to add data to the lockfile that it does not need.

Suppose I am barking down the path of build requirements of all dependencies not just the current project. One way to balance the performance concern would be to only do this when poetry is building from source, e.g., something like [`poetry config --local installer.no-binary :all:`](https://python-poetry.org/docs/configuration/#installerno-binary).

However, this config option only takes affect a _install time_ not _lock time_. So I was wondering if you would be open to accepting a PR that adds a no-binary option or flag to `poetry lock` and in that case specifically, puts the `build-system.requires` of every dependent packages into `poetry.lock` perhaps under a group, so that other users won't notice it.

I think this would be of use more generally than just for Poetry2Nix. Anyone who wants to do air-gapped builds (e.g., Spack, Bazel, or just machines without internet access for whatever reason) or build-from-source for security reasons would benefit from this.
